packages/testing/src/index.ts

```diff
@@ -0,0 +1,157 @@
+import * as activity from '@temporalio/activity';
+import { AsyncCompletionClient, Connection, WorkflowClient } from '@temporalio/client';
+import { ActivityFunction, CancelledFailure } from '@temporalio/common';
+import { Core } from '@temporalio/worker';
+import path from 'path';
+import { AbortController } from 'abort-controller';
+import { ChildProcess, spawn } from 'child_process';
+import events from 'events';
+import { kill, waitOnChild } from './child-process';
+import type getPortType from 'get-port';
+
+/**
+ * Options for {@link TestWorkflowEnvironment.create}
+ */
+interface TestWorkflowEnvironmentOptions {
+  testServerSpawner?(port: number): ChildProcess;
+}
+
+function defaultTestWorkflowEnvironmentOptions(): Required<TestWorkflowEnvironmentOptions> {
+  return {
+    testServerSpawner(port: number) {
+      return spawn(path.join(__dirname, '../test-server'), [`${port}`], {
+        stdio: 'ignore',
+      });
+    },
+  };
+}
+
+// TS transforms `import` statements into `require`s, this is a workaround until
+// tsconfig module nodenext is stable.
+const _importDynamic = new Function('modulePath', 'return import(modulePath)');
+
+/**
+ * An execution environment for running Workflow integration tests.
+ *
+ * Runs an external server.
+ * By default, the Java test server is used which supports time skipping.
+ */
+export class TestWorkflowEnvironment {
+  protected constructor(protected readonly serverProc: ChildProcess, protected readonly conn: Connection) {}
+
+  /**
+   * Create a new test environment
+   */
+  static async create(opts?: TestWorkflowEnvironmentOptions) {
+    // No, we're not going to compile this to ESM for one dependency
+    const getPort = (await _importDynamic('get-port')).default as typeof getPortType;
+    const port = await getPort();
+
+    const { testServerSpawner } = { ...defaultTestWorkflowEnvironmentOptions(), ...opts };
+
+    const child = testServerSpawner(port);
+
+    const address = `127.0.0.1:${port}`;
+    const conn = new Connection({ address });
+
+    try {
+      await Promise.race([
+        conn.untilReady(),
+        waitOnChild(child).then(() => {
+          throw new Error('Child exited prematurely');
+        }),
+      ]);
+    } catch (err) {
+      await kill(child);
+      throw err;
+    }
+
+    // TODO: Core is a singleton at the moment, once the bridge is refactored this will change.
+    await Core.install({ serverOptions: { address } });
+
+    return new this(child, conn);
+  }
+
+  /**
+   * Get an extablished {@link Connection} to the test server
+   */
+  get connection(): Connection {
+    return this.conn;
+  }
+
+  /**
+   * Get an {@link AsyncCompletionClient} for interacting with the test server
+   */
+  get asyncCompletionClient(): AsyncCompletionClient {
```

Comment:
One might expect a property to not construct a value (e.g. `env.asyncCompletionClient === env.asyncCompletionClient`). I think this should either be created once (eager or lazy) or be a function.

Reply:
👍

Reply:
Done